STUDENT PEROFORMANCE INDICATOR EXPLORATORY DATA ANALYSIS

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns 
import warnings
%matplotlib inline 
warnings.filterwarnings('ignore')

In [3]:
df=pd.read_csv('data/stud.csv')

In [4]:
df.head()

,gender,race_ethnicity,parental_level_of_education,lunch,test_preparation_course,math_score,reading_score,writing_score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75


In [5]:
df.shape

(1000, 8)

performing the datachecks

In [6]:
#checking the missing values 
df.isna().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

there are no missing values 

In [7]:
#checking duplicates
df.duplicated().sum()

np.int64(0)

there are no duplicates in the dataset 

In [8]:
#checking the datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race_ethnicity               1000 non-null   object
 2   parental_level_of_education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test_preparation_course      1000 non-null   object
 5   math_score                   1000 non-null   int64 
 6   reading_score                1000 non-null   int64 
 7   writing_score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [9]:
#checking the number of uniques values of each column 
df.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

In [10]:
df.describe()

,math_score,reading_score,writing_score
count,1000.00000,1000.000000,1000.000000
mean,66.08900,69.169000,68.054000
std,15.16308,14.600192,15.195657
min,0.00000,17.000000,10.000000
25%,57.00000,59.000000,57.750000
50%,66.00000,70.000000,69.000000
75%,77.00000,79.000000,79.000000
max,100.00000,100.000000,100.000000


from the above insights 
1. all means are very close to each other between 66 and 69.16
2. all standard deviations are also close between 14.6 and 15.19
3. for only maths there is an minimum score 0 for others minimum score is much higher

Exploring the data 

In [11]:
df.head()

,gender,race_ethnicity,parental_level_of_education,lunch,test_preparation_course,math_score,reading_score,writing_score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75


In [19]:
#checking the all the unique categories mentioned in each column 
print("categories in the gender column",df["gender"].unique())
print("categories in the race ethnicity",df["race_ethnicity"].unique())
print("categories in the parental level of education",df["parental_level_of_education"].unique())
print("categories in the lunch",df["lunch"].unique())
print("categories in the test prep course",df["test_preparation_course"].unique())


categories in the gender column ['female' 'male']
categories in the race ethnicity ['group B' 'group C' 'group A' 'group D' 'group E']
categories in the parental level of education ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
categories in the lunch ['standard' 'free/reduced']
categories in the test prep course ['none' 'completed']


In [31]:
#defining the numeric and the cateogrical features in the data 
numeric_features=[feature for feature in df.columns if df[feature].dtype!="O"]
categorical_features=[feature for feature in df.columns if df[feature].dtype=="O"]
print("No. of numeric features:", len(numeric_features),"->",numeric_features)
print("No. of categorical features:", len(categorical_features),"->",categorical_features)

No. of numeric features: 3 -> ['math_score', 'reading_score', 'writing_score']
No. of categorical features: 5 -> ['gender', 'race_ethnicity', 'parental_level_of_education', 'lunch', 'test_preparation_course']


adding the columns for total score and average score

In [35]:
df["total_score"]=df["math_score"]+df["reading_score"]+df["writing_score"]
df["avg_score"]=df["total_score"]/3
df.head(2)

,gender,race_ethnicity,parental_level_of_education,lunch,test_preparation_course,math_score,reading_score,writing_score,total_score,avg_score
0,female,group B,bachelor's degree,standard,none,72,72,74,218,72.666667
1,female,group C,some college,standard,completed,69,90,88,247,82.333333


In [43]:
#checking the number of students that got the full marks in respective subjects
math_full=df[df["math_score"]==100]["avg_score"].count()
reading_full=df[df["reading_score"]==100]["avg_score"].count()
writing_full=df[df["writing_score"]==100]["avg_score"].count()

print("No. of students with full score in maths:",math_full)
print("No. of students with full score in reading:",reading_full)
print("No. of students with full score in writing:",writing_full)

No. of students with full score in maths: 7
No. of students with full score in reading: 17
No. of students with full score in writing: 14
